### PR TITLE
refactor(programs): Consolidate Merkle trees into one PDA

### PR DESCRIPTION
**Pull request by vadorovsky**
Maintaining multiple PDAs requires more effort and we need to switch all types of Merkle trees at the same time anyway. We also need to make sure that the sizes of the Merkle trees are consistent. Keeping them in one PDA will make it easier.

Other changes:

* Merkle tree parameters
  * height: 22
  * roots: 2800

### CODE REVIEW

File: zk.js/src/merkleTree/merkleTreeConfig.ts [hunk 1]
```diff
@@ -68,29 +68,20 @@ export class MerkleTreeConfig {
...
+    const oldMerkleTreeSet = MerkleTreeConfig.getMerkleTreeSetPubkey(
+      merkleTreeSetIndex.sub(BN_1),
     );
-    const newEventMerkleTree =
-      MerkleTreeConfig.getEventMerkleTreePda(eventMerkleTreeIndex);
+    const newMerkleTreeSet =
+      MerkleTreeConfig.getMerkleTreeSetPubkey(merkleTreeSetIndex);
 
     const tx = await this.merkleTreeProgram.methods
-      .initializeNewMerkleTrees(DEFAULT_MERKLE_TREE_LOCK_DURATION)
```
Comment (ananas-block) [inline zk.js/src/merkleTree/merkleTreeConfig.ts:89]: Just making sure we don't have lock duration anywhere else anymore either right?

**Review by ananas-block — COMMENTED** [review]
Just two somewhat related comments.

File: zk.js/src/merkleTree/merkleTreeConfig.ts (hunk 1 above)
Comment (vadorovsky) [reply zk.js/src/merkleTree/merkleTreeConfig.ts:89]: Nope, after this PR we shouldn't :sweat_smile:

File: programs/merkle-tree/src/instructions/initialize_new_merkle_trees.rs [hunk 2]
```diff
@@ -61,33 +41,36 @@ pub fn process_initialize_new_merkle_trees(
...
-        new_transaction_merkle_tree,
-        merkle_tree_authority,
-        MERKLE_TREE_HEIGHT,
+    // Mark the old Merkle trees as not the newest
+    let loader: AccountLoader<MerkleTreeSet> = AccountLoader::try_from(
+        &ctx.remaining_accounts
+            .get(0)
+            .cloned()
+            .ok_or(ErrorCode::ExpectedOldMerkleTrees)?,
     )?;
```
[the remark this answers is not in the document]
Comment (ananas-block) [reply programs/merkle-tree/src/instructions/initialize_new_merkle_trees.rs]: yeah, I believe we were thinking that there should only be one tree at a time, but since that's obsolete we don't need it anymore.

File: zk.js/build-circuits/privateProgramTransaction4In4OutMain_js/witness_calculator.js [hunk 3]
```diff
@@ -0,0 +1,337 @@
+module.exports = async function builder(code, options) {
+
+    options = options || {};
```
[the remark this answers is not in the document]
Comment (ananas-block) [reply zk.js/build-circuits/privateProgramTransaction4In4OutMain_js/witness_calculator.js:3]: The build script should delete these but that doesn't work anymore apparently.

**Review by ananas-block — COMMENTED** [review]
Couple small things.

File: cli/package.json [hunk 4]
```diff
@@ -25,6 +25,7 @@
     "@oclif/plugin-not-found": "^2.4",
     "@oclif/plugin-plugins": "^3.8.4",
     "@project-serum/anchor": "^0.26.0",
+    "@solana-developers/helpers": "^1.4.2",
```
[the remark this answers is not in the document]
Comment (vadorovsky) [reply cli/package.json:28]: https://github.com/Lightprotocol/light-protocol/pull/444/files#diff-55e974c363e6ec4b6adf786e3612262c2d12eea6f243664f08fd44595ec57177R4

They provide a nice function for reading keypairs from files.

File: scripts/getAccountState.sh [hunk 5]
```diff
@@ -1,18 +1,17 @@
...
+    "merkle-tree-authority"
+    "merkle-tree-set-0"
+    "registered-spl-pool-token"
+    "registered-spl-pool"
+    "registered-sol-pool-token"
+    "registered-sol-pool"
+    "registered-verifier-0"
+    "registered-verifier-1"
+    "registered-verifier-2"
+    "registered-verifier-3"
```
[the remark this answers is not in the document]
Comment (vadorovsky) [reply scripts/getAccountState.sh:14]: Sounds good.

File: zk.js/build-circuits/privateProgramTransaction4In4OutMain_js/witness_calculator.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (vadorovsky) [reply zk.js/build-circuits/privateProgramTransaction4In4OutMain_js/witness_calculator.js:3]: The `rm` lines removing these files are commented :stuck_out_tongue: Should I uncomment?

File: programs/merkle-tree/src/verifier_invoked_instructions/insert_two_leaves_event.rs [hunk 6]
```diff
@@ -18,11 +15,8 @@ pub struct InsertTwoLeavesEvent<'info> {
         seeds::program = registered_verifier.pubkey,
     )]
     pub authority: Signer<'info>,
-    #[account(mut, seeds = [
-        EVENT_MERKLE_TREE_SEED,
-        event_merkle_tree.load().unwrap().merkle_tree_nr.to_le_bytes().as_ref()
-    ], bump)]
-    pub event_merkle_tree: AccountLoader<'info, EventMerkleTree>,
+    #[account(mut)]
+    pub merkle_tree_set: AccountLoader<'info, MerkleTreeSet>,
```
[the remark this answers is not in the document]
Comment (vadorovsky) [reply programs/merkle-tree/src/verifier_invoked_instructions/insert_two_leaves_event.rs:19]: It's a very good question and I was unsure myself.

I don't think there is any straightforward way of checking that MerkleTreeProgram owns the account? It's not PDA, so there is no `seed` or `seed::program`.

I think the best I can use is `address=` attribute. We could store a pubkey of the "newest"/"currently active" MTS in the MerkleTreeAuthority and base the `address` check on it. WDYT?

File: programs/merkle-tree/src/config_accounts/merkle_tree_authority.rs [hunk 7]
```diff
@@ -42,28 +31,8 @@ pub struct InitializeMerkleTreeAuthority<'info> {
...
-        seeds = [
-            EVENT_MERKLE_TREE_SEED,
-            0u64.to_le_bytes().as_ref(),
-        ],
-        bump,
-        payer=authority,
-        space = EventMerkleTree::LEN,
-    )]
-    pub event_merkle_tree: AccountLoader<'info, EventMerkleTree>,
+    #[account(zero)]
```
[the remark this answers is not in the document]
Comment (vadorovsky) [reply programs/merkle-tree/src/config_accounts/merkle_tree_authority.rs:34]: You don't need to mark the account as mutable in the init instruction (where you use `zero`), but you need to use `mut` in other instructions which use that (already initialized) account.

File: programs/merkle-tree/src/verifier_invoked_instructions/insert_two_leaves_event.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (vadorovsky) [reply programs/merkle-tree/src/verifier_invoked_instructions/insert_two_leaves_event.rs:19]: @ananas-block ping. Do you agree with this idea?

File: scripts/getAccountState.sh (hunk 5 above)
[the remark this answers is not in the document]
Comment (vadorovsky) [reply scripts/getAccountState.sh]: Done.

File: programs/merkle-tree/src/verifier_invoked_instructions/insert_two_leaves_event.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (ananas-block) [reply programs/merkle-tree/src/verifier_invoked_instructions/insert_two_leaves_event.rs:19]: tbh I think you cannot write into an account you don't own so ownership is checked with mut and probably also by account loader.

[the remark this answers is not in the document]
Comment (ananas-block) [reply programs/merkle-tree/src/verifier_invoked_instructions/insert_two_leaves_event.rs:19]: since we will have multiple Merkle trees the newest attribute doesn't make sense anymore.

File: verifier-sdk/src/light_transaction.rs [hunk 8]
```diff
@@ -156,12 +154,20 @@ impl<
...
+        custom_heap::log_total_heap("post load MerkleTreeSet");
+
+        // self.emit_indexer_transaction_event(&mut merkle_tree_set)?;
         self.emit_indexer_transaction_event()?;
         #[cfg(all(target_os = "solana", feature = "mem-profiling"))]
         custom_heap::log_total_heap("emit_indexer_transaction_event");
         self.compute_tx_integrity_hash()?;
         #[cfg(all(target_os = "solana", feature = "mem-profiling"))]
         custom_heap::log_total_heap("compute_tx_integrity_hash");
+        // self.fetch_root(&mut merkle_tree_set)?;
```
Comment (ananas-block) [inline verifier-sdk/src/light_transaction.rs:170]: leftover comment

File: zk.js/src/wallet/user.ts [hunk 9]
```diff
@@ -1610,10 +1591,10 @@ export class User {
   }) {
     const programAddress = getVerifierProgramId(idl);
     const balance = latestBalance
-      ? await this.syncStorage(idl, true, merkleTree)
+      ? await this.syncStorage(idl, true)
```
Comment (ananas-block) [inline zk.js/src/wallet/user.ts:1594]: Why did you remove this?
It doesn't matter because we are going to delete this code anyway, but in general on some rpc request level you will want the option to provide a Merkle tree set public key so that you define the search space you want to get your utxo from.

**Review by ananas-block — CHANGES_REQUESTED** [review]
Awesome that it's greeeen. 
Saw some leftover comments, and stuff in the user class but we will remove the user class anyway so don't bother changing that.

File: zk.js/src/wallet/user.ts (hunk 9 above)
Comment (vadorovsky) [reply zk.js/src/wallet/user.ts:1594]: Since `Rpc`/`TestRpc` is being passed to the `User` class, and it already contains the information about which MT is being used, passing it again in random parts of the class just seemed redundant for me. This way made fixing the merkle_tree_tests easier - I'm generating a new MTS keypair there and passing it once to TestRpc was just easier than passing it over in multiple places.

Before my changes, this `merkleTree` variable was set to `MerkleTreeConfig.getTransactionMerkleTreePda()` in some other random methods of `User` - `syncStorate`, `getUtxos` etc. I don't find this intuitive. I rather prefer to set the MT pubkey for the whole class once.

The other option would be to have it in the constructor.

Anyways, this is going to be removed, so maybe that doesn't matter.

File: verifier-sdk/src/light_transaction.rs (hunk 8 above)
Comment (vadorovsky) [reply verifier-sdk/src/light_transaction.rs]: Whoops, sorry.

Fixed. And I also added heap logs in the new places where I'm loading the MTS.

For more context - to get rid of "account already borrowed" error, I had to make sure that `merkle_tree_set` variables are dropped before doing CPI. Reading from MTS and then doing a CPI (with MTS as one of accounts) without dropping the MTS variable triggers the issue (which makes sense).

File: verifier-sdk/src/light_transaction.rs [hunk 10]
```diff
@@ -476,10 +475,16 @@ impl<
     }
 
     /// Fetches the root according to an index from the passed-in Merkle tree.
+    // pub fn fetch_root(&mut self, merkle_tree_set: &mut RefMut<'_, MerkleTreeSet>) -> Result<()> {
```
Comment (ananas-block) [inline verifier-sdk/src/light_transaction.rs:478]: one more leftover comment

**Review by ananas-block — APPROVED** [review]
One more leftover comment, otherwise looks good, feel free to merge once that's fixed.
